llvm/test/CodeGen/SPIRV/optimizations/add-check-overflow.ll

```diff
@@ -1,12 +1,14 @@
 ; This test aims to check ability to support "Arithmetic with Overflow" intrinsics
 ; in the special case when those intrinsics are being generated by the CodeGenPrepare;
-; pass during translations with optimization (note -O3 in llc arguments).
+; pass during translations with optimization (note -disable-lsr, to inhibit
+; strength reduction pre-empting with a more preferable match for this pattern
+; in llc arguments).
 
-; RUN: llc -O3 -mtriple=spirv32-unknown-unknown %s -o - | FileCheck %s
-; RUN: %if spirv-tools %{ llc -O3 -mtriple=spirv32-unknown-unknown %s -o - -filetype=obj | spirv-val %}
+; RUN: llc -O3 -disable-lsr -mtriple=spirv32-unknown-unknown %s -o - | FileCheck %s
```

Comment:
If the intent is to specifically check codegenprepare, should have an IR->IR test in test/Transforms/CodeGenPrepare.

I don't know whether the -disable-lsr output is the best or not, but based on the name of the test I would assume this would try to document the actual result, not with the special flag. 

Also this test shouldn't have been using -O3 (it barely does anything and -O2 is the default)

Reply:
I'm not entirely sure about the original intent. This retains the current behaviour of the test. Disabling LSR is necessary for that, and doing things like so doesn't harm coverage. I don't know that I want to go and fully re-develop this particular test at this point (it probably should be redone, I'm just not entirey keen since I'm not entirey sure what is desired), as part of this particular change. Doing it as strictly IR-to-IR would bypass the original intent of the SPIR-V folks here, which seems centered around lowering the staturated ops.

Reply:
Happy to help with motivation. The original intent is to have a minimized reproducer of an issue as a part of the test suite. The point of the test case is to feed SPIR-V Backend with LLVM IR without saturation arithmetic intrinsics and ensure that SPIR-V Backend is able to deal with those intrinsics appearing along the way in translation. There is a separate test case in the test suite for the same intrinsics explicitly represented in the input LLVM IR, and it's another case, different from this one.

Reply:
Well, I think that in this case @arsenm is on the money here: we should have an IR->IR test in CodeGenPrepare wherein we ensure that for the SPIR-V target, the optimisation triggers and the saturating intrinsics show up. Their correct lowering is, as you say, handled someplace else. Correctly selecting / lowering them should be independent of whether they were implicitly inserted by some optimisation pass, or explicitly inserted by the user.   Personally, I would separate that change from this PR (as it's pretty unrelated), and merely disable LSR in the existing test temporarily so as to not lose its coverage. Thoughts?

Reply:
In this case you lose coverage, because

> Correctly selecting / lowering them should be independent of whether they were implicitly inserted by some optimisation pass, or explicitly inserted by the user.

it's a nice idea, I agree with you, but it's somewhat detached. There are two different conditions: (1) those intrinsics are present already and explicitly accessible for processing by SPIRV Backend code, and (2) those intrinsics appear during later passes and not accessible for processing. This test case is related to the very concrete improvement.

Reply:
Thank you, I also agree that it doesn't seem appropriate to shift attention from the PR to the test case.

> I'm not exactly sure what loss of coverage you have in mind here
> what the very concrete improvement is in this case

As I've said, "you have explicitly inserted llvm intrinsics in LLVM IR code" and "you don't have it" means two different starting conditions for the backend, and different options on how to implement their translation. It's an implementation detail and is out of this PR's scope. Please feel free to ping me aside of this discussion in the case if you interested in the subject and would like to check implementation details and prior discussions.

Reply:
Thank you for clarifying, will definitely follow up, sounds interesting. Just to clarify, since it's probably not immediately apparent: without the patch, LSR wouldn't trigger for this loop, so disabling LSR, which can now optimise the loop, restores existing behaviour (it's not just a cheeky way to bypass the test).

Reply:
The purpose of this test appears to be to demonstrate the net result, which would be update (rather than disable lsr to get the previous output). Some other transform decided something else was better, should show what that is.

Reply:
I somehow get the sense that  this is not the desired outcome here, at least from the SPIR-V BE side, and we're cycling between these two states (this test has to absolutely stay as it is vs it should reflect reality). It would be nice to get a conclusion one way or the other so that this can be progressed / merged, unless there are other blocking concerns. The best I can think of as a solution is to just add an additional `RUN` with LSR enabled which reflects the net result?

Reply:
I think this should be updated to reflect the reality

Reply:
I've updated the test to reflect / test both the old output and what we currently get by default.